Added a new #SONGSELECT course tag

mjvotaw opened the pull request:
The current `#SONG` tag isn't very well documented, and it's capabilities are pretty limited. I want to introduce an alternative to it that is easier to use, and provides more flexible selection options. Because I couldn't think of a better name at the time, I've just called it `#SONGSELECT`. If anyone has a better idea, I'm very much open to changing it.

I've added a full description of the `#SONGSELECT` tag to Docs/CourseFormat.txt, but the basic idea is that you can compose a course entry by specifying one or more parameters, which follow the general format of
`#SONGSELECT:PARAM1=VALUE1:PARAM2=VALUE2:PARAM3=VALUE3;`

The available selection parameters are TITLE, GROUP, ARTIST, GENRE, DIFFICULTY, METER, BPMRANGE, DURATION, and SORT.

I also fixed a few issues that I came across while working on this, the two main ones were:
- CourseSortSongs() would cause a crash when leaving Marathon mode, but only when a course that uses SongSort_TopGrades or SongSort_LowestGrades was newly added (and it only happened on initial load because CourseWriterCRS didn't write Course caches correctly)
- Course entries that only specified a meter range would select different songs for different course difficulties.

Aside from that, and breaking out some of the course loading/writing logic into separate functions, I've avoided making any changes or optimizations that would affect how existing courses using the `#SONG` tag function.

mjvotaw committed:
Added BPMRANGE song option for Courses.
Started breaking out some of the logic in CourseLoaderCRS into smaller functions for readibility.

mjvotaw committed:
Changed SongCriteria.m_sGroupName to a vector m_vsGroupNames, to allow for the potential to select a song from more than one group.

mjvotaw committed:
Added additional criteria to StepsCriteria and SongCriteria
Added new m_sSongName property to Song class
Slight refactor in Song::LoadSongFromDir() to set m_sSongName
Implemented the basics of #SONGSELECT course parameter

mjvotaw committed:
Refactored code for writing #SONG entries in CourseWriterCRS
Added method for writing #SONGSELECT entries
Added new boolean flag bUseSongSelect to CourseEntry
Added new function StringToSongSort()
Renamed "LIVES" to "GAINLIVES"

mjvotaw committed:
Added OldStyleStringToSongSort() function to map best/worst/gradebest/gradeworst to their SongSort counterparts.
Added CourseLoaderCRS::SetCourseSongSort() to parse and set SONGSELECT's SORT parameter.

mjvotaw committed:
Added Artists, min/max bpm, min/max duration to CourseEntry::GetTextDescription()

mjvotaw committed:
Instead of calling rnd(), just select the average between iMaxDist and iMinDist.
This fixes an issue where a Course entry that only specifies a meter range will potentially select different songs for each difficulty.

mjvotaw committed:
Added description for the new #SONGSELECT course tag.

teejusb commented on Docs/CourseFormat.txt:
```diff
@@ -86,13 +86,149 @@ Complete Course Template
+--  - Unlike the #SONG tag, this will not take into account group names.
+--    If you want to specify a specific song from a specific group, use
+--    this along with the GROUP parameter.
+--  - This does not support names that have commmas.
```
If you add support for escape characters I think you should be able to do this. I suspect you'd also run into issues with titles that have # and ; in it as well.

mjvotaw replied:
This turned out to be slightly more complicated than I expected, but I just pushed some changes that will handle commas and control characters (`#`, `:`, `;`).

teejusb commented on Docs/CourseFormat.txt:
```diff
@@ -86,13 +86,149 @@ Complete Course Template
+#SONGSELECT:DIFFICULTY=Medium,Hard;
+
+--  METER= A minimum and maximum value of a song's difficulty meter,
+--  separated by EXACTLY TWO (2) periods.
```
Thoughts on using a dash? 8-12? Two periods is also fine, but was curious.

mjvotaw replied:
The two periods was just staying in line with how `#SONG` handled it, but using a dash is probably the more intuitive choise (I even messed this up last night while I was writing some courses to test this out, and kept putting dashes instead of `..`)

teejusb replied:
Ahhhh I didn't realize that's how #song did it. I wonder if just supporting both is feasible. Dash seems way more intuitive.

mjvotaw replied:
I'm don't think we need to support both, I'm just gonna go with dash.

teejusb commented:
Also I think this PR needs to be rebased (since it includes two commits that I dont think need to be included)

mjvotaw committed:
Uncommenting these lines from the original #SONG parsing, because I'm not sure why I commented it out to begin with.

mjvotaw committed:
Removed now un-used CourseLoaderCRS::ParseCourseSongSort
Replaced '..' with '-' for ranged parameter delimiter

mjvotaw committed:
Removed comments to myself

mjvotaw committed:
Updated CourseFormat.txt to replace the `..` with `-`

mjvotaw committed:
Default bUseSongSelect to false when initializing CourseEntry (otherwise it might be true, which causes CourseWriterCRS to write the cache incorrectly)

mjvotaw committed:
Modified MsdFile::ReadBuf() so that escaped characters are still skipped when bUnescape == false
As of this commit, this won't affect anything except CourseLoaderCRS, because everywhere else is passing `true` for the bUnescape parameter.

mjvotaw committed:
Added a new SmEscape parameter to allow us to specify what characters to escape (defaults to '\\', ':', ';')
Added new SmEscape function to handle escaping a vector of strings
Added SmUnescape() to remove the '\\' preceding an escaped character.

mjvotaw committed:
CourseWriterCRS now correctly escapes control characters for certain parameters
CourseLoaderCRS handles escaped control characters and commas
Added new parsing functions to CourseLoaderCRS, ParseCommaSeparatedList, and ParseRangedValue

mjvotaw committed:
Equals signs also need to be escaped when parsing #SONGSELECT parameters.

mjvotaw committed:
Removing already outdated comment

mjvotaw committed:
Explicitly get song bpm when checking if song matches songCriteria (m_fSpecifiedBPMMax and m_fSpecifiedBPMMin are only set if #DISPLAYBPM is used in simfile)

mjvotaw committed:
Updated notes on SORT parameter to indicate that it can actually be used with other parameters just fine, but the resutls can get kind of weird.

mjvotaw commented:
I think is probably as good as it's going to get for the moment. I've rebased the changes to remove the two commits I somehow picked up, fixed a few more song-selection issues, and done my best to make sure that none of this affects the selections made by the original `#SONG` tag.

teejusb commented:
I'm going to ahead and merge this since the next release is very soon. I haven't been able to extensively test this just due to lack of time but it seems like you've done your due diligence and at least the build is passing. We'll see if I hear any reports.
